File: rd-cpp/src/rd_framework_cpp/src/test/cases/RdAsyncSignalTest.cpp
#include "RdAsyncTestBase.h"

#include "RdSignal.h"
#include "Void.h"
#include "WireUtil.h"

#include "countdownlatch.hpp"

using namespace rd;
using namespace rd::test;
using namespace rd::test::util;


TEST_F(RdAsyncTestBase, DISABLED_asyncSignalStatic) {
	std::atomic_int32_t acc{0};

	clatch::countdownlatch evt1(3);
	clatch::countdownlatch evt2(1);

	RdSignal<Void> clientSignal;
	clientUiScheduler->queue([&] {
		statics(clientSignal, static_entity_id);
		bindStatic(clientProtocol.get(), clientSignal, static_name);
		clientSignal.async = true;

		evt1.count_down();
		evt1.await();
		clientBgScheduler->queue([&]() {
			clientSignal.fire();
			printf("client fired bg");
		});
		clientSignal.fire();
		printf("client fired ui");
		evt2.count_down();
	});

	RdSignal<Void> serverSignal;
	serverUiScheduler->queue([&]() {
		statics(serverSignal, static_entity_id);
		bindStatic(serverProtocol.get(), serverSignal, static_name);
		serverSignal.async = true;

		LifetimeDefinition::use([&](Lifetime lifetime) {
			serverSignal.advise_on(serverLifetime, serverBgScheduler.get(), [&] {
				printf("server received");
				rd::util::sleep_this_thread(100);
				serverBgScheduler->assert_thread();
				++acc;
			});
			printf("server advise completed");
			evt1.count_down();
			evt1.await();

		});
	});

	EXPECT_EQ(0, acc);
	evt1.count_down();
	evt1.await();


	evt2.await();
	clientUiScheduler->flush();
	clientBgScheduler->flush();

	serverUiScheduler->flush();
	serverBgScheduler->flush();

	EXPECT_EQ(2, acc);
	printf("end of all things");
}

File: rd-kt/rd-framework/src/test/kotlin/com/jetbrains/rd/framework/test/cases/RdAsyncSignalTest.kt
package com.jetbrains.rd.framework.test.cases

import com.jetbrains.rd.framework.base.static
import com.jetbrains.rd.framework.impl.RdSignal
import com.jetbrains.rd.framework.test.util.RdAsyncTestBase
import com.jetbrains.rd.util.lifetime.Lifetime
import com.jetbrains.rd.util.reactive.fire
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Test
import java.util.concurrent.CountDownLatch
import java.util.concurrent.atomic.AtomicInteger

class RdAsyncSignalTest : RdAsyncTestBase() {

    @Test
    fun testAsyncSignalStatic() {

        val acc = AtomicInteger(0)

        val evt1 = CountDownLatch(3)
        val evt2 = CountDownLatch(1)

        clientUiScheduler.queue {
            val clientSignal = clientProtocol.bindStatic(RdSignal<Unit>().static(1).apply { async = true }, "top")

            evt1.countDown()
            evt1.await()
            clientBgScheduler.queue {
                clientSignal.fire()
                println("client fired bg")
            }
            clientSignal.fire()
            println("client fired ui")
            evt2.countDown()
        }

        serverUiScheduler.queue {
            val serverSignal = serverProtocol.bindStatic(RdSignal<Unit>().static(1).apply { async = true }, "top")
            Lifetime.using {
                serverSignal.adviseOn(serverLifetime, serverBgScheduler) {
                    println("server received")
                    Thread.sleep(100)
                    serverBgScheduler.assertThread()
                    acc.incrementAndGet()
                }
                println("server advise completed")
                evt1.countDown()
                evt1.await()

            }
        }

        assertEquals(0, acc.get())
        evt1.countDown()
        evt1.await()

        evt2.await()
        clientUiScheduler.assertNoExceptions()
        clientBgScheduler.assertNoExceptions()


        serverUiScheduler.assertNoExceptions()
        serverBgScheduler.assertNoExceptions()

        assertEquals(2, acc.get())
        println("end of all things")
    }
}